src/envelope.rs

```diff
@@ -305,13 +277,30 @@ impl<D: Hash> Envelope<D> {
             .zip(self.ciphertext.iter())
             .map(|(&x1, &x2)| x1 ^ x2)
             .collect();
-        Ok(OpenedEnvelope {
+        Ok(OpenedInnerEnvelope {
             plaintext,
             export_key: *GenericArray::from_slice(&export_key),
         })
     }
 }
 
+// Helper functions
+
+fn construct_aad(server_s_pk: &[u8], optional_ids: &Option<(Vec<u8>, Vec<u8>)>) -> Vec<u8> {
+    let serialized_ids = match optional_ids {
```

Comment:
I believe this is `optional_ids.iter().flat_map(|(l, r)| [serialize(l, 2), serialize(r, 2)].concat()).collect()` [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=98ba4f0ce5c122d1d3034a8ff7488839)